src/main/java/org/opensearch/ad/ml/ModelManager.java

```diff
@@ -944,102 +944,121 @@ public void getTotalUpdates(String modelId, String detectorId, ActionListener<Lo
 
     /**
      * Compute anomaly result for the given data point
-     * @param detectorId Detector Id
      * @param datapoint Data point
-     * @param entityName entity's name like "server_1"
      * @param modelState the state associated with the entity
      * @param modelId the model Id
+     * @param detector Detector accessor
+     * @param entity entity accessor
+     *
      * @return anomaly result, confidence, and the corresponding RCF score.
      */
     public ThresholdingResult getAnomalyResultForEntity(
-        String detectorId,
         double[] datapoint,
-        String entityName,
         ModelState<EntityModel> modelState,
-        String modelId
+        String modelId,
+        AnomalyDetector detector,
+        Entity entity
     ) {
-        ThresholdingResult result = null;
-
         if (modelState != null) {
-            EntityModel model = modelState.getModel();
-            Queue<double[]> samples = model.getSamples();
-            samples.add(datapoint);
-            if (samples.size() > this.rcfNumMinSamples) {
-                samples.remove();
+            EntityModel entityModel = modelState.getModel();
+
+            if (entityModel == null) {
+                entityModel = new EntityModel(entity, new ArrayDeque<>(), null, null);
+                modelState.setModel(entityModel);
             }
 
-            result = maybeTrainBeforeScore(modelState, entityName);
+            if (entityModel.getRcf() == null || entityModel.getThreshold() == null) {
+                entityColdStarter.trainModelFromExistingSamples(modelState);
+            }
+
+            if (entityModel.getRcf() != null && entityModel.getThreshold() != null) {
+                return score(datapoint, modelId, modelState);
+            } else {
+                entityModel.addSample(datapoint);
+                return new ThresholdingResult(0, 0, 0);
+            }
         } else {
-            result = new ThresholdingResult(0, 0, 0);
+            return new ThresholdingResult(0, 0, 0);
         }
-
-        return result;
     }
 
-    private ThresholdingResult score(Queue<double[]> samples, String modelId, ModelState<EntityModel> modelState) {
+    public ThresholdingResult score(double[] feature, String modelId, ModelState<EntityModel> modelState) {
```

Comment:
Can you explain why this method changed? Before it was taking the last feature from `samples` to generate a result, now it's taking `feature` directly?

Reply:
Changed because I want the caller to tell me which point to score instead the method making an assumption that the last point is the point to return score for.